rfc/system/2633-eviction-of-cached-task-outputs.md

```diff
@@ -0,0 +1,125 @@
+# [RFC] Eviction of cached task outputs
+
+**Authors:**
+
+- @MorpheusXAUT
+- [blackshark.ai](https://blackshark.ai)
+
+This document: https://hackmd.io/qOztkaj4Rb6ypodvGEowAg?view
+
+## 1 Executive Summary
+
+This RFC describes an approach for cache eviction for Flyte executions, both on a workflow and task level. This will allow for invalidation of single (or multiple) task outputs for a single execution, providing an easy way to remove potentially invalid output data or cleaning up no longer required previous runs without having to change the `cache_version` (thus requiring a re-calculation of other related data).
+
+Additionally, we propose to expose the cache eviction functionality via `flyteadmin` as an API to clean up cached values of previously completed executions of workflows or tasks.
+
+## 2 Motivation
+
+Some large executions might produce invalid/undesirable results for some tasks - even though the execution doesn't fail, its output might still be faulty, requiring a re-calculation to be performed using identical inputs.
+
+At the moment, the only way to force Flyte to ignore the (faulty) cached value of a previous execution during a re-run of the workflow would be to change the `cache_version` of the affected task, causing all other calculations to be performed again as well even though their results might've been correct in the first place.
+
+Depending on the workload executed, this might result in unnecessary and potentially high costs and time requirements to re-process. Evicting some of the cached values would allow re-runs of previously executed workflows to leverage caching where desirable while still re-calculating some of the data that has been deemed "invalid".
+
+In addition, exposing cache eviction functionality via a `flyteadmin` API call would allow Flyte users/administrators to perform housekeeping and clean up their cache, cleanly removing the cached data of previously executions that are no longer relevant executions and would otherwise incur continuous storage costs. By exposing the same functionality used during the cache eviction override, we can ensure the cache maintained by `datacatalog` is consistent with the underlying blob storage.
```

Comment:
Whilst we didn't necessarily think about archive executions in the first place, you've definitely got a valid point there. I like the suggestion of optionally evicting the cache of tasks when archiving them, could also make it a default and add an opt-out flag (less clicking-effort when archiving multiple ones 😁 ).

We very rarely use the archiving function of Flyte internally. When talking about "no longer relevant" executions, I originally meant past executions that (while not being archived) are not relevant to any future calculations anymore because e.g. their input dataset turned out to be unusable.
At the moment, we're still keeping those around unarchived because their workflow/task is still relevant, however the output of certain executions is no longer of interest to us and basically just eats up blob storage space.

I can add cache eviction on archival to the draft as I believe it could be a useful addition that should be very little effort to include from the beginning, thanks for the suggestion, didn't think about that 🙂

Reply:
Adding @katrogan